File: src/lua/utils/run_all_graphs.lua
local loader = require('src.lua.utils.load_graph')
local h_bt = require('src.lua.backtrack.backtrack')
local h_dp = require('src.lua.dynamic.dynamic')
local h_nn = require('src.lua.heuristic.heuristic')
local decorators = require('src.lua.utils.decorators')

local function run_algorithms_on_file(filepath)
  print('\n' .. string.rep('=', 70))
  print('Archivo:', filepath)
  print(string.rep('=', 70))

  local ok, graph = pcall(loader.load_graph_txt, filepath)
  if not ok then
    print('  -> Error al cargar el grafo:', graph)
    return
  end

  local function run_and_print(algorithm_func, label)
    print('\n[' .. label .. ']')

    local context = decorators.new_execution_context()
    local path, cost = algorithm_func(graph, context)

    print(string.format('--- Ejecutando función: %s ---', label))

    if path then
      print('  -> Ruta encontrada:', table.concat(path, ' -> '))
      if cost == math.huge then
        print('  -> Costo del circuito: N/A')
      else
        print(string.format('  -> Costo del circuito: %.2f', cost))
      end
    else
      print('  -> No se encontró un circuito hamiltoniano.')
    end

    print(string.format('  -> Tiempo: %.4fs', context.time_seconds or 0))
    print(string.format('  -> Memoria estimada usada: %.2f KB',
                        context.memory_peak_kb or 0))
    print(string.format('  -> Operaciones: %s', context.ops_count or 0))
    print(string.rep('-', 40))
  end

  run_and_print(h_bt, 'Backtracking')
  if #graph <= 20 then
    run_and_print(h_dp, 'Programación Dinámica')
  else
    print('  -> Programación Dinámica omitida por tamaño del grafo > 20')
  end
  run_and_print(h_nn, 'Heurística - Vecino Más Cercano')
end

local function get_txt_files_from_dir(path)
  local handle = io.popen('find "' .. path .. '" -type f -name "*.txt"')
  local result = {}
  if handle then
    for line in handle:lines() do
      table.insert(result, line)
    end
    handle:close()
  end
  return result
end

local files = get_txt_files_from_dir('data')
for _, filepath in ipairs(files) do
  run_algorithms_on_file(filepath)
end

File: src/python/utils/run_all_graphs.py
import os

from src.python.backtrack.backtrack import hamiltonian_backtracking
from src.python.dynamic.dynamic import hamiltonian_dp
from src.python.heuristic.heuristic import hamiltonian_nearest_neighbor
from src.python.utils.decorators import ExecutionContext
from src.python.utils.load_graph import load_graph_txt

INF = float('inf')


def run_algorithms_on_file(filepath):
  print('\n' + '=' * 70)
  print(f'Archivo: {filepath}')
  print('=' * 70)

  try:
    graph = load_graph_txt(filepath)
  except Exception as e:
    print(f'  -> Error al cargar el grafo: {e}')
    return

  def run_and_print(algorithm_func, label):
    print(f'\n[{label}]')
    context = ExecutionContext()
    path, cost = algorithm_func(graph, context=context)
    print(f'--- Ejecutando función: {algorithm_func.__name__} ---')
    if path:
      print(f'  -> Ruta encontrada: {' -> '.join(map(str, path))}')
      if cost == INF:
        print('  -> Costo del circuito: N/A')
      else:
        print(f'  -> Costo del circuito: {cost:.2f}')
    else:
      print('  -> No se encontró un circuito hamiltoniano.')

    # Mostrar métricas del contexto
    print(f'  -> Tiempo: {context.time_seconds:.4f}s')
    print(f'  -> Memoria máxima: {context.memory_peak_kb:.2f} KB')
    print(f'  -> Operaciones: {context.ops_count:,}')
    print('-' * 40)

  run_and_print(hamiltonian_backtracking, 'Backtracking')
  if len(graph) <= 20:
    run_and_print(hamiltonian_dp, 'Programación Dinámica')
  else:
    print('  -> Programación Dinámica omitida por tamaño del grafo > 20')
  run_and_print(hamiltonian_nearest_neighbor, 'Heurística - Vecino Más Cercano')


def run_on_all_txt_files(root_folder='data'):
  for subdir, _, files in os.walk(root_folder):
    for file in files:
      if file.endswith('.txt'):
        filepath = os.path.join(subdir, file)
        run_algorithms_on_file(filepath)


if __name__ == '__main__':
  run_on_all_txt_files('data')
